'actAsRegularWindow' contributor setting implementation

darksworm opened the pull request:
<!--

Have you read Code of Conduct? By filing an Pull Request, you are expected to comply with it, including treating everyone with respect: https://github.com/hovancik/stretchly/blob/master/CODE_OF_CONDUCT.md

-->

Issue: #437
<!-- Link to relevant issue. All PRs should be associated with an issue -->

### Requirements

* Filling out the template is required. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.

<!-- If there’s a checkbox you can’t complete for any reason, that's okay, just explain in detail why you weren’t able to do so. -->

- [x]  issue was opened to discuss proposed changes before starting implementation. It is important do discuss changes before implementing them (Why should we add it? How should it work? How should it look? Where will it be? ...).
- [x]  during development, `node` version specified in `package.json` was used (ie using [nvm](https://github.com/creationix/nvm)).
- [x]  package versions and package-lock.json were not changed (`npm install --no-save`).
- [x]  app version number was not changed.
- [ ]  all new code has tests to ensure against regressions. - :exclamation: **I'm afraid that the only way to test this reasonably is via acceptance tests and as they are all commented out (`test/main.js`) I haven't put in the effort**
- [x] `npm run lint` reports no offenses.
- [ ] `npm run test` is error-free. <details><summary>:exclamation: **unrelated tests are failing** </summary> ![image](https://user-images.githubusercontent.com/9987548/109414560-29fa1580-79bc-11eb-839f-ec96e7af2d9a.png)</details>
- [x]  README and CHANGELOG were updated accordingly.
- [x]  after PR is approved, all commits in it are [squashed](https://gitbetter.substack.com/p/how-to-squash-git-commits)

### Description of the Change

Added "Make stretchly act as a regular window" setting to the contributor settings. When enabled, you can move, minimize and send-to-back the stretchly break window. You cannot however close the window.
<!--

We must be able to understand the design of your change from this description. If we can't get a good idea of what the code will be doing from the description here, the pull request may be closed at the maintainers' discretion. Keep in mind that the maintainer reviewing this PR may not be familiar with or have worked with the code here recently, so please walk us through the concepts.

-->


### Verification Process

<!--

What process did you follow to verify that your change has the desired effects?

- How did you verify that all new functionality works as expected?
- How did you verify that all changed functionality works as expected?
- How did you verify that the change has not introduced any regressions?

Describe the actions you performed (e.g., buttons you clicked, text you typed, commands you ran, etc.), and describe the results you observed.

-->
### Test scenario:
1. set "Show breaks in" to "Window"
2. set "Make stretchly act as a regular window" to ON :heavy_check_mark: 
3. "Skip to next" -> "Mini break" via icon in corner of taskbar

#### Results:
-> Window has title bar, break window appears in taskbar, window cannot be closed via ALT-F4 or exit button, window can be minimized, showAsRegularWindow setting appears in contributor settings.

#### Repeat same for "Long break":
-> Same results as previous.

#### Repeat same for "Show breaks in" = "Fullscreen" with both Mini and Long breaks:
->  Window covering whole screen appears, has a title bar, appears in task bar, cannot be closed via ALT-F4 or exit button, window can be minimized.

#### Repeat same for "Long break", same results.

<details>
<summary>Testing on windows - proof</summary>

https://user-images.githubusercontent.com/9987548/109429519-bf20fc80-7a04-11eb-994b-5f6f1bd0d0a0.mp4


https://user-images.githubusercontent.com/9987548/109429523-bfb99300-7a04-11eb-8d34-1e4145bbfc7f.mp4


https://user-images.githubusercontent.com/9987548/109429524-c0522980-7a04-11eb-88da-97f7c10629c6.mp4


https://user-images.githubusercontent.com/9987548/109429525-c0522980-7a04-11eb-9d45-32e55df6b496.mp4


https://user-images.githubusercontent.com/9987548/109429597-2343c080-7a05-11eb-868b-09b05890c3b0.mp4

</details>

<details>
<summary>Testing on linux - proof</summary>

Super+Q is the hotkey to close a window.


https://user-images.githubusercontent.com/9987548/109429994-96e6cd00-7a07-11eb-99ab-b286440def7e.mp4


https://user-images.githubusercontent.com/9987548/109429995-977f6380-7a07-11eb-9549-489a47c2b954.mp4


https://user-images.githubusercontent.com/9987548/109429996-977f6380-7a07-11eb-81bc-81665d54662b.mp4


https://user-images.githubusercontent.com/9987548/109429997-9817fa00-7a07-11eb-9b89-17a8071f643a.mp4


https://user-images.githubusercontent.com/9987548/109430001-9e0ddb00-7a07-11eb-89b9-648ceda47df7.mp4


https://user-images.githubusercontent.com/9987548/109430002-9ea67180-7a07-11eb-8233-e505669dc6a2.mp4

</details>


### Other information

I have no means of testing how this behaves on macos - help with that would be appreciated

darksworm committed:
basic 'actAsRegularWindow' contributor setting implementation https://github.com/hovancik/stretchly/issues/437

darksworm committed:
move act as regular window option above theme section

darksworm committed:
missing const

darksworm commented:
I'm having troubles testing this on windows, when I build it for windows with `electron-builder --x64 --win` and launch it, I get these errors:

![image](https://user-images.githubusercontent.com/9987548/109414924-76465500-79be-11eb-8a9d-64485562cf39.png)

Googling these has not been very helpful - seems to be something to do with having 32 bit depedencies? IDK.

@hovancik have you experienced this issue?

hovancik commented:
what's your OS, 32 or 64?

darksworm commented:
> what's your OS, 32 or 64?

64bit

darksworm commented:
> > what's your OS, 32 or 64?
> 
> 64bit

tried building on windows - that solved the issue.

darksworm committed:
hr should appear on new line

darksworm committed:
add new setting description in changelog and readme

darksworm committed:
capitalize Stretchly

darksworm commented:
@hovancik this is ready for review.

hovancik commented on README.md:
```diff
@@ -173,6 +173,10 @@ To specify the size of the break window, set the value of `breakWindowHeight` an
 
+#### Make Stretchly act as a regular window [![Contributor Preferences](https://img.shields.io/badge/Contributor_Preferences-✔-success)](#contributor-preferences)
+
+If you want Stretchly to act as a regular window (have a titlebar, turn off always on top, be minimizable) set `actAsRegularWindow` to `false`.
```
this should be `true`?

darksworm replied:
yup, nice catch!

hovancik commented on app/main.js:
```diff
@@ -471,8 +472,11 @@ function startMicrobreak () {
-    microbreakWinLocal.setAlwaysOnTop(true, 'screen-saver')
+    microbreakWinLocal.setAlwaysOnTop(actAsRegularWindow, 'screen-saver')
     if (microbreakWinLocal) {
+      microbreakWinLocal.on('close', (e) => {
```
What is this doing?

darksworm replied:
This prevents the break window from being closed.

darksworm replied:
because:
- allowing the window to be `focusable` lets the users ALT-F4, CTRL-W to close it
- adding the window to the taskbar lets the users close the window by right-clicking it in the taskbar
- adding the titlebar lets the users press the cross on the window to close it

AFAIK there is no way to remove those close buttons in electron besides removing the frame (which I've avoided to let users drag and minimize the window in this mode)

hovancik replied:
Ah, ok :) Makes sense

hovancik commented on app/contributor-preferences.html:
```diff
@@ -411,6 +411,11 @@
+        <input type="checkbox" class="enabletype" value="actAsRegularWindow" id="actAsRegularWindow">
+        <label data-i18next="contributorPreferences.actAsRegularWindow" for="actAsRegularWindow"></label>
+      </div>
+      <div></div>
```
Let's make changes here: 
![image](https://user-images.githubusercontent.com/4745859/109636746-3f05ae80-7b4c-11eb-8ceb-bed39a0456d0.png)

1. Let's do it similar to this: 
![image](https://user-images.githubusercontent.com/4745859/109636831-5b095000-7b4c-11eb-8acc-49932cddfd13.png)

For example: 
```
Act as regular window :  When enabled, Break window will become focusable, it won't be 'Always on top' anymore and it will become available in taskbar. 

(checkbox) Show breaks as regular windows
``` 

You are also missing translations. You only need to add English ones. 

You can test this window by changing `isContributor: false` to `isContributor: true` in `main.js`

darksworm replied:
changed in https://github.com/hovancik/stretchly/pull/875/commits/404d234471252d8e005ec6bae18c706e599efd81

![image](https://user-images.githubusercontent.com/9987548/109698776-70a46700-7b98-11eb-8ae7-a73a64455fb1.png)

hovancik replied:
Looks good! 

Small issue with next block; `Appearance` should be on left side

![image](https://user-images.githubusercontent.com/4745859/109813883-32598700-7c2e-11eb-8bf4-4eb26508c9f4.png)

darksworm replied:
Don't know how I missed that!
Fixed in https://github.com/hovancik/stretchly/pull/875/commits/b31b5a70433819498f2ba88fc6f124807d64dfca
![image](https://user-images.githubusercontent.com/9987548/109825732-c46b8a80-7c42-11eb-9dbd-f52e9858f79f.png)

hovancik commented:
I am trying on Ubuntu 20.04 and when Break is on, the window is not gone when I click on other windows. 

Is that expected? I thought the point was to be able to go to another windows.

darksworm committed:
fix wrong boolean value in readme

darksworm committed:
set alwaysontop to false when acting as regular window

darksworm commented:
> I am trying on Ubuntu 20.04 and when Break is on, the window is not gone when I click on other windows.
> 
> Is that expected? I thought the point was to be able to go to another windows.

no it is not, the boolean value for the second set of `setAlwaysOnTop` calls was inversed. fixed.

darksworm committed:
rename showBreaksAsRegularWindows, put it in a seprate section in preferences, update readme

darksworm commented:
@hovancik this is ready for your review

hovancik commented:
Haven't tested elsewhere, but on my Ubuntu, break is not opened correctly: I do not see the break window, I only see new item in tasks list: 

![image](https://user-images.githubusercontent.com/4745859/109814223-8c5a4c80-7c2e-11eb-9ff5-e25e8dc4e53d.png)

hovancik commented:
Can you also try quitting the app from tray while break is on? It seems like whole app kinda freezes.

darksworm committed:
fix contributor settings title alignment

darksworm committed:
focus break window if it is displayed as regular window

darksworm commented:
> Haven't tested elsewhere, but on my Ubuntu, break is not opened correctly: I do not see the break window, I only see new item in tasks list:
> 
> ![image](https://user-images.githubusercontent.com/4745859/109814223-8c5a4c80-7c2e-11eb-9ff5-e25e8dc4e53d.png)

Couldn't replicate on windows or DWM on linux, but managed to replicate this on budgie - it was due to the window not being focused upon creation, i guess different window managers handle this differently. 

fixed in https://github.com/hovancik/stretchly/pull/875/commits/3f7d5ec8b05b0ee14a2af68ab0b5d29a1bff975d (also noticed the macos kiosk mode settings, adapted those as well to work in line with the new setting)

darksworm committed:
allow strecthly to be closed while break windows open

darksworm commented:
> Can you also try quitting the app from tray while break is on? It seems like whole app kinda freezes.

Yup! Turns out the code which stopped the break windows from being closed also stopped the whole application from being closed while they were visible.

I've changed it so that when calling `app.quit`, the break windows will close.

---
I've addressed all issues, waiting for further comments.

hovancik commented on app/main.js:
```diff
@@ -450,10 +455,15 @@ function startMicrobreak () {
+      }
+
       log.info(`Stretchly: showing window ${localDisplayId + 1} of ${numberOfDisplays()}`)
       if (process.platform === 'darwin') {
```
Did you test on macOS, then? Elsewhere you wrote you can't test there

darksworm replied:
No, I did not. Sadly, I don't have access to a macOS device.

hovancik replied:
Fullscreen does not seems to work, will try investigate later

hovancik replied:
I think it should be 
```
      if (process.platform === 'darwin') {
        if (showBreaksAsRegularWindows) {
          microbreakWinLocal.setFullScreen(settings.get('fullscreen'))          
        } else {
          microbreakWinLocal.setKiosk(settings.get('fullscreen'))
        }
      }
```

darksworm replied:
Yup! That makes a lot of sense to me - I've applied those changes here - https://github.com/hovancik/stretchly/pull/875/commits/a66893e72dc6d7115bfebb52f0ab4b4f8a5e4db0

darksworm committed:
wip: fixing e2e tests

darksworm committed:
basic e2e tests

darksworm committed:
reformat file

darksworm committed:
test microbreak window when `showBreaksAsRegularWindows` enabled

darksworm committed:
make e2e tests run only on macos

darksworm commented:
I've rewritten the end-to-end tests (`main.js`) to make them work and added partial coverage for this feature.

I cannot however get the tests to run on windows and ubuntu in the github action, so i've changed the CI to launch the end-to-end tests only on macOS.

Although not great, it does seem better than not running them at all. They won't be included in the coverage, but I feel that's OK - otherwise e2e tests generate fake coverage anyway.

darksworm committed:
don't fail workflow if windows job fails

hovancik commented:
Thanks for the tests! 
What was the error? I believe I disabled the tests because of this: https://github.com/electron-userland/spectron/issues/656

darksworm commented:
> Thanks for the tests!
> What was the error? I believe I disabled the tests because of this: [electron-userland/spectron#656](https://github.com/electron-userland/spectron/issues/656)

Yeah, i encountered that error - it seems that the "promiseness" transfer just doesn't work - I rewrote the tests to just use `await` instead.

---

The e2e tests fail on github actions with this currently:
```
unknown error: Chrome failed to start: exited abnormally.
  (unknown error: DevToolsActivePort file doesn't exist)
  (The process started from chrome location /home/runner/work/stretchly/stretchly/node_modules/spectron/lib/launcher.js is no longer running, so ChromeDriver is assuming that Chrome has crashed.)
      at Object.startWebDriverSession (node_modules/webdriver/build/utils.js:34:15)
      at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

Most articles suggest adding some chrome driver options: `--headless,--no-sandbox,--disable-gpu,--disable-dev-shm-usage` but these make no difference.

Also tried setting up a virtual display with `xvfb` - didn't help either.

The tests run just fine on my local linux machine - that makes me believe this has to be be something to do with the headless github actions environment.

darksworm committed:
fix fullscreen for mac when showing breaks as regular windows

hovancik reviewed (approved):
I guess we are ready to merge, then?

hovancik commented on app/main.js:
```diff
@@ -471,8 +485,13 @@ function startMicrobreak () {
-    microbreakWinLocal.setAlwaysOnTop(true, 'screen-saver')
+    microbreakWinLocal.setAlwaysOnTop(!showBreaksAsRegularWindows, 'screen-saver')
     if (microbreakWinLocal) {
+      microbreakWinLocal.on('close', (e) => {
```
So looks like this is causing issues on macOS: #903
